File: apps/routes/promo.py
from flask import Blueprint, request

from apps.services import Promo
from apps.services.BaseServices import token_auth

promo = Blueprint("promo", __name__, url_prefix="/api/promo")
promo.before_request(lambda: token_auth.login_required(lambda: None)())

@promo.route("", methods=["GET"])
def _getPromo():
    filters = request.args.to_dict()

    if filters:
        return Promo().getPromoFiltered(filters)
    else:
        return Promo().getPromo()

@promo.route("/<string:kode_promo>", methods=["GET"])
def _getPromoByKode(kode_promo): return Promo().getPromoByKode(kode_promo)

@promo.route("/dropdown-data", methods=["GET"])
def _getDropdownData(): return Promo().getDropdownData()

@promo.route("/dropdown-data-log", methods=["GET"])
def _getDropdownDataLog(): return Promo().getDropdownDataLog()

@promo.route("/dropdown-data-klaim", methods=["GET"])
def _getDropdownDataKlaim(): return Promo().getDropdownDataKlaim()

@promo.route("/generate-code", methods=["POST"])
def _generateCode(): return Promo().generateKlaimCode()

@promo.route("/list-faktur", methods=["GET"])
def _getListFaktur(): return Promo().getListFaktur()

@promo.route("/klaim-kategori", methods=["GET"])
def _getKlaimKategori(): return Promo().getKlaimKategori()

@promo.route("/ajukan-klaim", methods=["POST"])
def _ajukanDataKlaim(): return Promo().ajukanDataKlaim()

@promo.route("/ajukan-ulang-klaim", methods=["PUT"])
def _ajukanUlangKlaim(): return Promo().ajukanUlangKlaim()

@promo.route("/log-penggunaan", methods=["GET"])
def _getLogPenggunaan(): return Promo().getLogPenggunaanPromo()

@promo.route("/klaim-promo", methods=["GET"])
def _getKlaimPromo():
    filters = request.args.to_dict()
    return Promo().getKlaimPromo(filters)

@promo.route("/klaim-promo/ditolak", methods=["GET"])
def _getKlaimPromoDitolak():
    filters = request.args.to_dict()
    filters['status_klaim'] = '3'
    return Promo().getKlaimPromo(filters)

@promo.route("/klaim-promo/detail/<int:id>", methods=["GET"])
def _getKlaimPromoDetail(id):
    return Promo().getKlaimDetail(id)

@promo.route("/kasbon-klaim", methods=["GET"])
def _getKasbonKlaim():
    filters = request.args.to_dict()
    return Promo().getKasbonKlaim(filters)

@promo.route("/kasbon-klaim/ajukan", methods=["POST"])
def _ajukanKasbonKlaim():
    return Promo().ajukanKasbonKlaim()

@promo.route('/klaim-promo/update-status/<int:klaim_id>', methods=['PUT'])
def update_klaim_status(klaim_id):
    data = request.get_json() or {}
    status = data.get('status_klaim')
    return Promo().updateKlaimStatus(klaim_id, status)

@promo.route("/kasbon-klaim/<int:id_kasbon_klaim>", methods=["GET"])
def _getKasbonKlaimDetail(id_kasbon_klaim):
    return Promo().getKasbonKlaimDetail(id_kasbon_klaim)

@promo.route("/kasbon-klaim/<int:id_kasbon_klaim>/klaim", methods=["GET"])
def _getKlaimForKasbonKlaim(id_kasbon_klaim, filters=None):
    filters = request.args.to_dict()
    return Promo().getListDetailKasbonKlaim(id_kasbon_klaim, filters)

@promo.route("/kasbon-klaim/<int:id_kasbon_klaim>/konfirmasi", methods=["POST"])
def _konfirmasiKasbonKlaim(id_kasbon_klaim):
    return Promo().konfirmasiKasbonKlaim(id_kasbon_klaim)

File: 12budimas-promo-dan-klaim/src/services/promo.js
import { fetchWithAuth } from "../lib/utils";
import { baseService } from "./baseService";

class PromoService extends baseService {
    constructor() {
        super();
        this.endpoints = {
            promo: `/api/promo`,
            promoListFaktur: `/api/promo/list-faktur`,
            generateKlaimCode: `/api/promo/generate-code`,
            klaimKategori: `/api/promo/klaim-kategori`,
            logPenggunaanPromo: `/api/promo/log-penggunaan`,
            klaimPromo: `/api/promo/klaim-promo`,
            kasbonKlaim: `/api/promo/kasbon-klaim`,
        };
        this.setServiceName("promo");
    }

    _buildUrlWithParams(baseUrl, params = {}) {
        const queryParams = new URLSearchParams();
        Object.keys(params).forEach(key => {
            if (params[key] !== undefined && params[key] !== null) {
                queryParams.append(key, params[key]);
            }
        });
        return queryParams.toString() ? `${baseUrl}?${queryParams.toString()}` : baseUrl;
    }

    _handleError(error, errorMessage) {
        console.log("error in service : ", error);
        this.throwError(`${errorMessage}: ${error.message}`);
        throw error;
    }

    async _fetch(method, url, payload = null, { wrapInArray = false, errorMessage } = {}) {
        try {
            const res = await fetchWithAuth(method, url, payload);
            return wrapInArray ? (Array.isArray(res) ? res : [res]) : res;
        } catch (error) {
            this._handleError(error, errorMessage);
        }
    }

    async getDropdownData() {
        const url = `${this.endpoints.promo}/dropdown-data`;
        return this._fetch("GET", url, null, {
            wrapInArray: true,
            errorMessage: "error while fetching promo dropdown data"
        });
    }

    async getDropdownDataLog() {
        const url = `${this.endpoints.promo}/dropdown-data-log`;
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching log dropdown data"
        });
    }

    async getDropdownDataKlaim(status_klaim = null) {
        let conditions = status_klaim ? `?status_klaim=${status_klaim}` : "";
        const url = `${this.endpoints.promo}/dropdown-data-klaim${conditions}`;
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching klaim dropdown data"
        });
    }

    async generateKlaimCode(payload) {
        const url = this.endpoints.generateKlaimCode;
        return this._fetch("POST", url, payload, {
            errorMessage: "error while generating klaim code"
        });
    }

    async updateKlaimStatus(payload) {
        const url = `${this.endpoints.klaimPromo}/update-status/${payload.id}`;
        return this._fetch("PUT", url, payload, {
            errorMessage: "error while updating klaim status"
        });
    }

    async getListFaktur(params = {}) {
        const baseUrl = this.endpoints.promoListFaktur;
        const url = this._buildUrlWithParams(baseUrl, params);
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching faktur list"
        });
    }

    async getKlaimKategori() {
        const url = this.endpoints.klaimKategori;
        return this._fetch("GET", url, null, {
            wrapInArray: true,
            errorMessage: "error while fetching klaim kategori"
        });
    }

    async getLogPenggunaanPromo(params = {}) {
        const baseUrl = this.endpoints.logPenggunaanPromo;
        const url = this._buildUrlWithParams(baseUrl, params);
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching log penggunaan promo"
        });
    }

    async getListKlaimPromo(params = {}) {
        const baseUrl = this.endpoints.klaimPromo;
        const url = this._buildUrlWithParams(baseUrl, params);
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching klaim promo list"
        });
    }

    async getListKasbonKlaim(params = {}) {
        const baseUrl = this.endpoints.kasbonKlaim;
        const url = this._buildUrlWithParams(baseUrl, params);
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching kasbon klaim list"
        });
    }

    async getKasbonKlaimDetail(id) {
        const url = `${this.endpoints.kasbonKlaim}/${id}`;
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching kasbon klaim detail"
        });
    }

    async getListKlaimForKasbonKlaim(id, params = {}) {
        const baseUrl = `${this.endpoints.kasbonKlaim}/${id}/klaim`;
        const url = this._buildUrlWithParams(baseUrl, params);
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching klaim for kasbon klaim"
        });
    }

    async getDetailPromo({ kode_promo } = {}) {
        const url = `${this.endpoints.promo}/${kode_promo}`;
        return this._fetch("GET", url, null, {
            wrapInArray: true,
            errorMessage: "error while fetching detail promo"
        });
    }

    async getKlaimDetail(id) {
        const url = `${this.endpoints.klaimPromo}/detail/${id}`;
        return this._fetch("GET", url, null, {
            errorMessage: "error while fetching klaim detail by kode"
        });
    }

    async postAjukanKlaim(payload) {
        const url = `${this.endpoints.promo}/ajukan-klaim`;
        return this._fetch("POST", url, payload, {
            errorMessage: "error while submitting klaim"
        });
    }

    async ajukanUlangKlaimData(payload) {
        const url = `${this.endpoints.promo}/ajukan-ulang-klaim`;
        return this._fetch("PUT", url, payload, {
            errorMessage: "error while submitting ulang klaim"
        });
    }

    async postAjukanKasbonKlaim(payload) {
        const url = `${this.endpoints.kasbonKlaim}/ajukan`;
        return this._fetch("POST", url, payload, {
            errorMessage: "error while submitting kasbon klaim"
        });
    }

    async konfirmasiDetailKasbonKlaim(id_kasbon_klaim, payload) {
        const url = `${this.endpoints.kasbonKlaim}/${id_kasbon_klaim}/konfirmasi`;
        return this._fetch("POST", url, payload, {
            errorMessage: "error while confirming kasbon klaim detail"
        });
    }
}

const promoService = new PromoService();
export { promoService };